GVFS/GVFS.Common/Git/LibGit2RepoInvoker.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Threading;
+
+namespace GVFS.Common.Git
+{
+    public class LibGit2RepoInvoker : IDisposable
+    {
+        private static readonly TimeSpan DefaultRepositoryDisposalPeriod = TimeSpan.FromMinutes(15);
+
+        private readonly Func<LibGit2Repo> createRepo;
+        private readonly ITracer tracer;
+        private readonly object sharedRepoLock = new object();
+        private readonly TimeSpan sharedRepositoryDisposalPeriod;
+        private volatile bool stopped;
+        private volatile int activeCallers;
+        private LibGit2Repo sharedRepo;
+        private Timer sharedRepoDisposalTimer;
+
+        public LibGit2RepoInvoker(ITracer tracer, Func<LibGit2Repo> createRepo, TimeSpan? disposalPeriod = null)
+        {
+            this.tracer = tracer;
+            this.createRepo = createRepo;
+
+            if (!disposalPeriod.HasValue || disposalPeriod.Value <= TimeSpan.Zero)
+            {
+                this.sharedRepositoryDisposalPeriod = DefaultRepositoryDisposalPeriod;
+            }
+            else
+            {
+                this.sharedRepositoryDisposalPeriod = disposalPeriod.Value;
+            }
+
+            this.sharedRepoDisposalTimer = new Timer(

```

Comment:
There is a 15 minute cleanup window here.

What happens if Expire background job runs before the dispose?  Would it be successful but potentially leave a .idx (or many)?

Reply:
We could leave .idx files on disk if libgit2 has a handle. That gets reported as part of the logging for that step.